keras/backend/torch/layer.py

```diff
@@ -13,8 +13,8 @@ def _post_build(self):
         self._track_variables()
 
     def _track_variables(self):
-        self.torch_params = torch.nn.ParameterList(
-            [variable.value for variable in self.variables]
+        self.torch_params = torch.nn.ParameterDict(
```

Comment:
The dict is more readable, but it's unsafe -- there's no guarantee that variable names are unique within a model (except for Functional models). It's entirely possible to create models with duplicate variable paths, which would cause tracking issues above. So the list is preferable.

Reply:
It should be safe to use `id(variable)` as a key for `torch.nn.ParameterDict`. I believe `BaseOptimizer` adopts the same approach to get the mapping of variables.

I could not find a solution for safely adding/removing the variable from `torch.nn.ParameterList`
(`in` and `remove` of `KerasVariable` are not supported)

Reply:
What's the original issue with using a list though?

Reply:
```python
from keras import layers

layer = layers.Dense(units=16)
layer.build([None, 8])
assert len(layer.torch_params) == len(layer.variables)

layer.enable_lora(rank=2)
assert len(layer.torch_params) == len(layer.variables)  # <--

```
|`master`|this pr|
|-|-|
|Failed at line 8|Success|

The biggest issue is that `zero_grad()` will fail to reset the uncaptured variables.
I believe the currect LoRA implementation will not be trained correctly on torch backend.

Reply:
To demonstrate the error of `zero_grad` using LoRA:
```python
import torch

from keras import layers

layer = layers.Dense(units=3)
layer.build([None, 2])
layer.enable_lora(rank=1)


def list_grads(layer):
    grads = dict()
    for v in layer.trainable_variables:
        grads[v.name] = v.value.grad
    return grads


# Fake grads
for v in layer.trainable_variables:
    v.value.grad = torch.rand_like(v.value)

print(list_grads(layer))
layer.zero_grad()
print(list_grads(layer))

```

```bash
# master branch
{'bias': tensor([0.6259, 0.4827, 0.6012], device='cuda:0'), 'lora_kernel_a': tensor([[0.6620],
        [0.7231]], device='cuda:0'), 'lora_kernel_b': tensor([[0.7123, 0.9257, 0.1676]], device='cuda:0')}
{'bias': None, 'lora_kernel_a': tensor([[0.6620],
        [0.7231]], device='cuda:0'), 'lora_kernel_b': tensor([[0.7123, 0.9257, 0.1676]], device='cuda:0')}

# this pr
{'bias': tensor([0.5960, 0.2336, 0.1569], device='cuda:0'), 'lora_kernel_a': tensor([[0.9123],
        [0.9217]], device='cuda:0'), 'lora_kernel_b': tensor([[0.3435, 0.9276, 0.6599]], device='cuda:0')}
{'bias': None, 'lora_kernel_a': None, 'lora_kernel_b': None}
```